### DISCUSSION TITLE
creation an XML file of server

### DISCUSSION THREAD

**Discussion by AndreyKl4** [FreeOpcUa/python-opcua#1245] [Q&A]
Hello all!

I really new with python opc ua. In my project I need to get a server interface in one XML file and after that I want to download this interface in Siemens OPC UA Client. 
Can smb help me pls in what way I can get an XML file of Server? I tried to do it with XmlExporter but i didn't manage to create a one xml file with all nodes, all data types, all custom data types in one file...

Thanks in advance!

**Comment by AndreasHeine** [comment]
Hey, the Server Class has the two methods:
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/server/server.py#L520
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/server/server.py#L528

typcially a OPC UA Client dont need the Servers xml models because they can browse the adressspace and load custom types!
i would assume that you mean something different (if you are new to opcua) so maybe tell a little more about what do you want to do!?

**Reply by AndreyKl4** [reply]
@AndreasHeine Thanks for the quick answer!
In TIA Portal in OPC UA Client interface there are 2 opportunities to get xml models from server:
1) to get interface automatically(online).
2) to download interface manually.

I tried the first variant and it works fine but only with simple data types(not custom). In my project i have a huge number of custom data types. I have monitored some forums and people advised to create server interface as xml file and after that to download  it manually. After these manipulations the problem with "not understanding custom structures"  (on siemes opc ua client side) should be solved.

**Reply by AndreasHeine** [reply]
i am assuming python-opcua is server and the s7 is client and you use the siemens opcua client "all in one" functionblock (Read/Write/Call)?
did you have your servers-interface in a seperat namespace? (`server.register_namespace("http://my-awsome-namespace/UA")`)
if yes! you simply can "export_xml_by_ns" means export by namespace
 """
        Export nodes of one or more namespaces to an XML file.
        Namespaces used by nodes are always exported for consistency.
        Args:
            server: opc ua server to use
            path: name of the xml file to write
            namespaces: list of string uris or int indexes of the namespace to export, if not provide all ns are used except 0
        Returns:
"""

edit:
but there is also the possibility to make a read and writelist manually in TIA Portal if i remember right you just need to stick to siemes DB structure!

**Comment by AndreyKl4** [comment]
@AndreasHeine Thanks a lot. I managed to create a xml file! 
You are right there is a possibility but i have a huge number of custom types and to define them manually it is a bit problematic...

**Reply by AndreyKl4** [reply]
@AndreasHeine One more question please, is it possible to get not only all nodes of the server but also data type dictionionary in one xml? I have tried to create server-interface for Siemens Client with help SiOME and i got the xml with all nodes and data type dictionary. After that i create xml of my python-server with help above mentioded method with same nodes(same structure) but xml looks different... Xml from SioMe is compatible with Siemens opc ua client but XML from opcua-python not... 
I think the problem is: i don't transfer DataType DIctionary things....

I attached both xmls...


[python_opcua.txt](https://github.com/FreeOpcUa/python-opcua/files/5931425/python_opcua.txt)
[siome.txt](https://github.com/FreeOpcUa/python-opcua/files/5931426/siome.txt)


Thanks in Advance!

**Reply by AndreasHeine** [reply]
why you dont list them on your own and export the nodes with: 
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/server/server.py#L520

while modelling the server addreassspace you can simply apend the nodes to a list or so and then export them

if that does not work:
contact Siemens Support... i am not sure if all of that is supported, they had trouble with some external and custom typedefinitions! but i am not up to date on that topic...
